test/widgets/action_sheet_test.dart

```diff
@@ -99,46 +107,244 @@ void main() {
+                of: find.byType(ReactionButtons),
+                matching: find.byType(InkWell)),
+              matching: find.byType(UnicodeEmojiWidget)),
+            matching: find.text(emojiDisplay.emojiUnicode)));
+        }
+
+        testWidgets('${emoji.emojiName} adding success', (tester) async {
+          final message = eg.streamMessage();
+          await setupToMessageActionSheet(tester, message: message, narrow: TopicNarrow.ofMessage(message));
+
+          connection.prepare(json: {});
+          await tapButton(tester);
+          await tester.pump(Duration.zero);
+
+          check(connection.lastRequest).isA<http.Request>()
+            ..method.equals('POST')
+            ..url.path.equals('/api/v1/messages/${message.id}/reactions')
+            ..bodyFields.deepEquals({
+                'reaction_type': 'unicode_emoji',
+                'emoji_code': emoji.emojiCode,
+                'emoji_name': emoji.emojiName,
+              });
+        });
+
+        testWidgets('${emoji.emojiName} removing success', (tester) async {
+          final message = eg.streamMessage(
+            reactions: [Reaction(
+              emojiName: emoji.emojiName,
+              emojiCode: emoji.emojiCode,
+              reactionType: ReactionType.unicodeEmoji,
+              userId: eg.selfAccount.userId)]
+          );
+          await setupToMessageActionSheet(tester, message: message, narrow: TopicNarrow.ofMessage(message));
+
+          connection.prepare(json: {});
+          await tapButton(tester);
+          await tester.pump(Duration.zero);
+
+          check(connection.lastRequest).isA<http.Request>()
+            ..method.equals('DELETE')
+            ..url.path.equals('/api/v1/messages/${message.id}/reactions')
+            ..bodyFields.deepEquals({
+                'reaction_type': 'unicode_emoji',
+                'emoji_code': emoji.emojiCode,
+                'emoji_name': emoji.emojiName,
+              });
+        });
+
+        testWidgets('${emoji.emojiName} request has an error', (tester) async {
+          final message = eg.streamMessage();
+          await setupToMessageActionSheet(tester, message: message, narrow: TopicNarrow.ofMessage(message));
+
+          connection.prepare(httpStatus: 400, json: {
+            'code': 'BAD_REQUEST',
+            'msg': 'Invalid message(s)',
+            'result': 'error',
           });
+          await tapButton(tester);
+          await tester.pump(Duration.zero); // error arrives; error dialog shows
+
+          await tester.tap(find.byWidget(checkErrorDialog(tester,
+            expectedTitle: 'Adding reaction failed',
+            expectedMessage: 'Invalid message(s)')));
+        });
+      }
     });
 
-    testWidgets('request has an error', (tester) async {
-      final message = eg.streamMessage();
-      await setupToMessageActionSheet(tester, message: message, narrow: TopicNarrow.ofMessage(message));
+    group('emoji picker;', () {
```

Comment:
nit: the code this is really testing is in `lib/widgets/emoji_reaction.dart`, so let's put the test in the corresponding file `test/widgets/emoji_reaction_test.dart`

(I see it's using the `setupToMessageActionSheet` helper; but it can import that from this file)